### PR TITLE
CDI-643 Provide a way to easily configure injection point of an InjectionTarget

### PR DESCRIPTION
**Pull request by antoinesd**
CDI-643 Provide a way to easily configure injection point of an InjectionTarget

### CODE REVIEW

**Comment by antoinesd** [comment]
Spec part not written yet. API should be discussed first.

File: api/src/main/java/javax/enterprise/inject/spi/InjectionTargetFactory.java [hunk 1]
```diff
@@ -56,4 +59,19 @@
...
+     * {@link #createInjectionTarget(Bean)} will use the {@link AnnotatedTypeConfigurator} to define injection points in the returned
+     * {@link InjectionTarget}
+     *
+     * Each call returns the same AnnotatedTypeConfigurator.
+     *
+     * @return an {@link AnnotatedTypeConfigurator} to configure injection points
+     * @since 2.0
+     */
+    default AnnotatedTypeConfigurator<T> configure() {
+        return null;
```
Comment (manovotn) [inline api/src/main/java/javax/enterprise/inject/spi/InjectionTargetFactory.java]: I don't think this is good. In fact, returning `null` in default impl seems rather nasty.
I can understand that you meant to provide a default method, however this class is not implemented by end user but rather by the CDI impl itself, therefore *not having* it `default` wouldn't cause so much harm IMO.

Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/InjectionTargetFactory.java]: I also think that the default impl is not necessary here.

Comment (antoinesd) [inline api/src/main/java/javax/enterprise/inject/spi/InjectionTargetFactory.java]: I agree that it's not nice, but look at the javadoc of the interface. It shows creation of an anonymous `InjectionTargetFactory` to wrap the one returned by the `BeanManager`. This code will break if user did that and update to 2.0...

Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/InjectionTargetFactory.java]: Ok, then throwing `UnsupportedOperationException` might be more appropriate...

**Comment by manovotn** [comment]
LGTM, +1

**Comment by mkouba** [comment]
Hm, previously `getInjectionTargetFactory()` returned a fully configured (possibly immutable) instance. If we allow to configure the annotated type, then we must specify when exactly is the factory configuration "finished" (effectively immutable). I.e. when `createInjectionTarget()` is called for the first time. And possibly throw `IllegalStateException` for subsequent ` configure()` calls?

File: spec/src/main/asciidoc/core/spi.asciidoc [hunk 2]
```diff
@@ -579,11 +579,17 @@ public <T> InjectionTargetFactory<T> getInjectionTargetFactory(AnnotatedType<T>
 public interface InjectionTargetFactory<T> {
 
     public InjectionTarget<T> createInjectionTarget(Bean<T> bean);
+    AnnotatedTypeConfigurator<T> configure();
```
Comment (mkouba) [inline spec/src/main/asciidoc/core/spi.asciidoc]: **public** AnnotatedTypeConfigurator<T> configure();

**Comment by mkouba** [comment]
Well, it might make sense to have the same text in javadoc and spec. And to be honest, I had to re-read the javadoc several times to understand it. But otherwise looks good ;-)

**Comment by antoinesd** [comment]
@mkouba, I aligned first sentence of javadoc on spec text..

**Comment by johnament** [comment]
This change doesn't make much sense to me.  The use case is that I can implement my own `InjectionTarget` and use it here.  This code has me potentially implementing `configure` as well here.  Are you expecting end users to implement `AnnotatedTypeConfigurator` here?

**Comment by antoinesd** [comment]
@johnament, nobody implements `InjectionTargetFactory` from scratch. You do that to decorate the one retuned by the `BeanManager` as shown in the Javadoc of the class.
As `BeanManager.createInjectionTarget()` is deprecated a common use is to use the factory to make a non contextual object an injection target.
Adding this configuration allows easy addition of injection points in the `InjectionTarget` to create.

**Comment by manovotn** [comment]
I am fine with this PR, do we have any objections or can we merge it?

**Comment by antoinesd** [comment]
I've created a very basic example to use this feature: https://gist.github.com/antoinesd/c51cdbfa4348abaeacc2024551f86e8d

**Pull request merged**